src/store.c

```diff
@@ -68,42 +71,82 @@ struct rill_store
 
     struct header *head;
     struct vals *vals;
-    uint8_t *data;
-    struct index *index;
+
+    uint8_t *data_a;
+    uint8_t *data_b;
+    struct index *index_a;
+    struct index *index_b;
     uint8_t *end;
 };
 
+struct rill_space
+{
+    size_t header_bytes;
+    size_t index_bytes[2];
+    size_t pairs_bytes[2];
+};
+
 
 // -----------------------------------------------------------------------------
 // coder
 // -----------------------------------------------------------------------------
 
 static struct encoder store_encoder(
-        struct rill_store *store, struct indexer *indexer)
+        struct rill_store *store,
+        struct indexer *indexer,
+        struct vals* vals,
+        uint64_t offset)
 {
     return make_encoder(
-            store->vma + store->head->data_off,
-            store->vma + store->vma_len,
-            store->vals, indexer);
-}
-
-static struct decoder store_decoder(struct rill_store *store)
-{
-    return make_decoder(
-            store->vma + store->head->data_off,
+            store->vma + offset,
             store->vma + store->vma_len,
-            store->vals, store->index);
+            vals,
+            indexer);
 }
 
 static struct decoder store_decoder_at(
-        struct rill_store *store, size_t key_idx, uint64_t off)
+        struct rill_store *store,
+        size_t key_idx,
+        uint64_t curr_off,
+        enum rill_col column)
 {
+    size_t offset = 0;
+    size_t offset_end = 0;
+    struct index* lookup = NULL;
+    struct index* index = NULL;
+
+    switch (column) {
+    case rill_col_a:
+        lookup = store->index_b;
+        index  = store->index_a;
+        offset = store->head->data_a_off;
+        offset_end = store->head->data_b_off;
+        break;
+    case rill_col_b:
+        lookup = store->index_a;
+        index  = store->index_b;
+        offset = store->head->data_b_off;
+        offset_end = store->vma_len;
+        break;
+    default:
+        rill_fail("improper rill col passed");
```

Comment:
I recommend including contextual information in your messages to make it easier to debug. In this case, I would include the value of `column` in the error message as the first thing I would do if that error message popped-up is to look for the value of that variable.